src/Combinatorics/IntegerCombinatorics/types.jl

```diff
@@ -0,0 +1,120 @@
+################################################################################
+#
+#  Partition
+#
+################################################################################
+
+@doc raw"""
+    Partition{T<:IntegerUnion} <: AbstractVector{T}
+
+A **partition** of a non-negative integer ``n`` is a decreasing sequence ``λ₁ ≥ λ₂ ≥ … ≥
+λᵣ`` of positive integers ``λᵢ`` such that ``n = λ₁ + … + λᵣ``. The ``λᵢ`` are called the
+**parts** of the partition and ``r`` is called the **length**.
+
+A partition can be encoded as an array with elements ``λᵢ``. We provide the parametric type
+`Partition{T}` which is a subtype of `AbstractVector{T}` where `T` can be any subtype of
+`IntegerUnion`. All functions that can be used for vectors (1-dimensional arrays) can thus
+be used for partitions as well. There is no performance impact by using an own type for
+partitions rather than simply using arrays. The parametric type allows to increase
+performance by using smaller integer types. For efficiency, the `Partition` constructor does
+not check whether the given array is indeed a decreasing sequence.
+
+A partition can be created by either calling `partition` on an array of integers or by
+calling `partition` with arguments being the sequence of parts, with the possibility
+to provide the element type as the first argument.
+
+# Examples
+```jldoctest
+julia> P = partition([6,4,4,2]) #The partition 6+4+4+2 of 16.
+[6, 4, 4, 2]
+
+julia> P = partition(6,4,4,2) #Same as above but less to type
+[6, 4, 4, 2]
+
+julia> length(P)
+4
+
+julia> P[1]
+6
+```
+Usually, ``|λ| ≔ n`` is called the **size** of ``λ``. In Julia, the function `size` for
+arrays already exists and returns the *dimension* of an array. Instead, you can use the
+Julia function `sum` to get the sum of the parts.
+```jldoctest
+julia> P = partition(6,4,4,2)
+[6, 4, 4, 2]
+
+julia> sum(P)
+16
+```
+You can create partitions with smaller integer types as follows.
+```jldoctest
+julia> P = partition(Int8,6,4,4,2) #Or partition(Int8[6,4,4,2])
+Int8[6, 4, 4, 2]
+```
+There is a unique partition of 0, namely the **empty partition** (of length 0). It can be
+created as follows.
+```jldoctest
+julia> P = partition() #Or partition([])
+Int64[]
+julia> sum(P)
+0
+julia> length(P)
+0
+julia> P = partition(Int8) #Or partition(Int8[])
+Int8[]
+```
+
+# References
+1. [Ful97](@cite)
+2. [Knu11](@cite), Section 7.2.1.4 (starting on page 390).
+"""
+struct Partition{T<:IntegerUnion} <: AbstractVector{T}
+  p::Vector{T}
+end
+
+################################################################################
+#
+#  Tableau
+#
+################################################################################
+
+@doc raw"""
+    Tableau{T} <: AbstractVector{AbstractVector{T}}
```

Comment:
stupid question maybe, but why don't we make the name more explicit? Even the docstring only talks about "Young tableaus".
```suggestion
    YoungTableau{T} <: AbstractVector{AbstractVector{T}}
```

Reply:
I don't have an opinion (I don't work with tableaux). In AbstractAlgebra it is called `YoungTableau` too. Either way, there should be a lower case constructor (`tableau` or `young_tableau` or both) which is so far missing.

Reply:
Maybe we can use `YoungTableau` to be on the safe side for now. We could change it later. The other way around would be more difficult.